### PR TITLE
Generate enum variant assist

### LINKED ISSUE #11635: New assist: add enum variant
image

It'd be cool if in this situation rust-analyzer realized that that what we want here is not a function, but a new error variant. 

Looking at how generate function works should help with implementing this assist: 

https://github.com/rust-analyzer/rust-analyzer/blob/5b9e46b960c37f04a26742c186e0a26b7bc131b3/crates/ide_assists/src/handlers/generate_function.rs

Note that generate function grew quite complex over time, for learning purposes, you might want to check some of its earlier versions

### PR DESCRIPTION
**Pull request by fasterthanlime**
So, this is kind of a weird PR!

I'm a complete newcomer to the `rust-analyzer` codebase, and so I browsed the "good first issue" tag, and found #11635. Then I found two separate folks had taken stabs at it, most recently @maartenflippo — and there had been a review 3 days ago, but no activity in a little while, and the PR needed to be rebased since the crates were renamed from `snake_case` to `kebab-case`.

So to get acquainted with the codebase I typed this PR by hand, looking at the diff in #11995, and I also added a doc-test (that passes).

I haven't taken into account the comments @Veykril left in #11995, but I don't want to steal any of @maartenflippo's thunder! Closing this PR is perfectly fine. Or Maarten could use it as a "restart point"? Or I could finish it up, whichever feels best to everyone.

I think what remains to be done in this PR, at least, is:

  * [x] Only disable the "generate function" assist if the name is `PascalCase`
  * [x] Only enable the "generate variant" assistant if the name is `PascalCase`
  * [x] Simplify with `adt.source()` as mentioned here: https://github.com/rust-lang/rust-analyzer/pull/11995#discussion_r875134175
  * [ ] Add more tests for edge cases? Are there cases where simply adding one more indent level than the enum's indent level is not good enough? Some nested trickery I'm not thinking of right now?

Anyway. This PR can go in any direction. You can tell me "no, tackle your own issue!" And I'll go do that and still be happy I got to take a look at rust-analyzer some by doing this. Or you can tell me "okay, now _you_ finish it", and I guess I'll try and finish it :)

Closes #11635

### CODE REVIEW

File: crates/ide-assists/src/handlers/generate_enum_variant.rs [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    let path_expr: ast::PathExpr = ctx.find_node_at_offset()?;
+    let path = path_expr.path()?;
+
+    if ctx.sema.resolve_path(&path).is_some() {
+        // No need to generate anything if the path resolves
+        return None;
+    }
+
+    let name_ref = path.segment()?.name_ref()?;
+    if name_ref.text().as_str().chars().next()?.is_ascii_lowercase() {
```
Comment (Veykril) [inline crates/ide-assists/src/handlers/generate_enum_variant.rs:44]:
```suggestion
    if name_ref.text().as_str().starts_with(char::is_lowercase) {
```
Similar can be done for the uppercase check in `generate_function`. Going with the non-ascii version probably makes more more sense? (though I don't really know of anyone NOT using the latin alphabet set for identifiers)

**Comment by Veykril** [comment]
> Or you can tell me "okay, now you finish it", and I guess I'll try and finish it :)

I suppose there is no need to say this anymore 😄

Since you have included the original commit author for the rebased commit I personally think this is fine.

You are more than welcome to pick up your own issues from here on of course (and ask for guidance/mentoring on some that seem interesting to you :) )

@bors delegate+

File: crates/ide-assists/src/handlers/generate_enum_variant.rs [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        format!("{}", IndentLevel(1))
+    };
+    let text = format!("{}{},\n{}", prefix, name_ref, enum_indent_level);
+
+    acc.add(
+        AssistId("generate_enum_variant", AssistKind::Generate),
+        "Generate variant",
+        target,
+        |builder| builder.insert(offset, text),
+    )
```
Comment (Veykril) [inline crates/ide-assists/src/handlers/generate_enum_variant.rs:84]: Ah, missed this. Let's pull the `text` creation into the closure since that is evaluated lazily.

File: crates/ide-assists/src/handlers/generate_enum_variant.rs (hunk 1 above)
Comment (fasterthanlime) [reply crates/ide-assists/src/handlers/generate_enum_variant.rs]: Done in 796c4d8a10359bc8add6e5acf952c87f4354189b

File: crates/ide-assists/src/handlers/generate_enum_variant.rs (hunk 2 above)
Comment (fasterthanlime) [reply crates/ide-assists/src/handlers/generate_enum_variant.rs]: Done in ae2c0db67f0408128500dd80c1a7b1e9a118ac54

**Comment by fasterthanlime** [comment]
Took care of those last two review comments — the only thing I'm still unhappy with it the indentation in the method generation assist, but I suppose that's for a separate PR:

https://github.com/rust-lang/rust-analyzer/pull/12334/files#diff-e36280c9e19413d3eab3f92468e2cd3a9798708843c568a8b6f7a184b5d6c129R1763-R1783

For now I just made the test match the assist's output.

**Comment by Veykril** [comment]
Ye, indentation and whitespace formatting is a big annoyance for us in assists still. We try to fix that up to some degree manually (within reasonable limits of code readability). So having non-perfect output for these is alright for the time being. We are planning on having a proper formatter that works on our syntax trees in the future to get rid of all those manual fixups.

Thanks!
@bors r+

**Comment by lnicola** [comment]
https://user-images.githubusercontent.com/308347/169745420-959a2079-4123-4a52-8ccd-bd1386783706.mp4
